### PR TITLE
UAR-1508 Ensure BOs don't lose their trust associations when updated

### PR DESCRIPTION
**Pull request by mwestacott**
### JIRA link

https://companieshouse.atlassian.net/browse/UAR-1508

### Change description

* Trust ids associated with BOs now persist, even after BO details are edited
* Works also for BOs been reviewed on the Update/Remove journeys
* Unit tests updated to relect and check the change in behaviour

### Work checklist

- [X] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

**Review by markpit — APPROVED** [review]

File: src/controllers/update/update.review.beneficial.owner.individual.ts [hunk 1]
```diff
@@ -55,21 +55,32 @@ export const post = async (req: Request, res: Response, next: NextFunction) => {
...
-      const haveDayOfBirth = appData.beneficial_owners_individual[Number(boiIndex)].have_day_of_birth;
+    if (boiIndex !== undefined && appData.beneficial_owners_individual && appData.beneficial_owners_individual[Number(boiIndex)].id === req.body["id"]) {
+      const boData: BeneficialOwnerIndividual = appData.beneficial_owners_individual[Number(boiIndex)];
+      const boId = boData.id;
+      const dob = boData.date_of_birth as InputDate;
+      const haveDayOfBirth = boData.have_day_of_birth;
+
+      let trustIds: string[] = [];
+      if (boData.trust_ids) {
+        trustIds = boData.trust_ids;
```
Comment (lduranteau) [inline src/controllers/update/update.review.beneficial.owner.individual.ts:66]: Use the spread operator to recreate a new array or else they will have the same reference [...boData.trust_ids]

Your if condition doesn't check  if the array has a size if (boData?.trust_ids.length);

You can do everything in one line:
const trustIds: string[] = boData?.trust_ids.length ? [...boData.trust_ids] : [];

File: src/controllers/update/update.review.beneficial.owner.individual.ts [hunk 2]
```diff
@@ -55,21 +55,32 @@ export const post = async (req: Request, res: Response, next: NextFunction) => {
...
+
       const session = req.session as Session;
 
       const data: ApplicationDataType = setBeneficialOwnerData(req.body, uuidv4());
       if (haveDayOfBirth) {
         data[HaveDayOfBirthKey] = haveDayOfBirth;
       }
 
+      if (trustIds.length > 0) {
+        (data as BeneficialOwnerIndividual).trust_ids = trustIds;
```
Comment (lduranteau) [inline src/controllers/update/update.review.beneficial.owner.individual.ts:81]: Same for the reference

File: src/controllers/update/update.review.beneficial.owner.other.ts [hunk 3]
```diff
@@ -63,14 +63,25 @@ export const post = async (req: Request, res: Response, next: NextFunction) => {
...
     const appData = getApplicationData(req.session);
 
-    if (booIndex !== undefined && appData.beneficial_owners_corporate && appData.beneficial_owners_corporate[Number(booIndex)].id === req.body["id"]){
-      const boId = appData.beneficial_owners_corporate[Number(booIndex)].id;
+    if (booIndex !== undefined && appData.beneficial_owners_corporate && appData.beneficial_owners_corporate[Number(booIndex)].id === req.body["id"]) {
+      const boData: BeneficialOwnerOther = appData.beneficial_owners_corporate[Number(booIndex)];
+      const boId = boData.id;
+
+      let trustIds: string[] = [];
+      if (boData.trust_ids) {
```
Comment (lduranteau) [inline src/controllers/update/update.review.beneficial.owner.other.ts:71]: same as above

File: src/controllers/update/update.review.beneficial.owner.other.ts [hunk 4]
```diff
@@ -63,14 +63,25 @@ export const post = async (req: Request, res: Response, next: NextFunction) => {
...
+      }
+
       removeFromApplicationData(req, BeneficialOwnerOtherKey, boId);
 
       const session = req.session as Session;
 
       const data: ApplicationDataType = setBeneficialOwnerData(req.body, uuidv4());
 
+      if (trustIds.length > 0) {
+        (data as BeneficialOwnerOther).trust_ids = trustIds;
```
Comment (lduranteau) [inline src/controllers/update/update.review.beneficial.owner.other.ts:82]: Same as above

File: src/utils/beneficial.owner.individual.ts [hunk 5]
```diff
@@ -119,11 +120,24 @@ export const updateBeneficialOwnerIndividual = async (req: Request, res: Respons
   try {
     logger.debugRequest(req, `UPDATE ${req.route.path}`);
 
+    const id = req.params[ID];
+    const boData: BeneficialOwnerIndividual = getFromApplicationData(req, BeneficialOwnerIndividualKey, id, true);
+
+    let trustIds: string[] = [];
```
Comment (lduranteau) [inline src/utils/beneficial.owner.individual.ts:126]: Same as above

File: src/utils/beneficial.owner.individual.ts [hunk 6]
```diff
@@ -119,11 +120,24 @@ export const updateBeneficialOwnerIndividual = async (req: Request, res: Respons
...
+
     // Remove old Beneficial Owner
-    removeFromApplicationData(req, BeneficialOwnerIndividualKey, req.params[ID]);
+    removeFromApplicationData(req, BeneficialOwnerIndividualKey, id);
 
     // Set Beneficial Owner data
-    const data: ApplicationDataType = setBeneficialOwnerData(req.body, req.params[ID]);
+    const data: ApplicationDataType = setBeneficialOwnerData(req.body, id);
+
+    if (trustIds.length > 0) {
```
Comment (lduranteau) [inline src/utils/beneficial.owner.individual.ts:137]: Same as above

File: src/utils/beneficial.owner.other.ts [hunk 7]
```diff
@@ -104,11 +104,23 @@ export const updateBeneficialOwnerOther = async (req: Request, res: Response, ne
   try {
     logger.debugRequest(req, `UPDATE ${req.route.path}`);
 
+    const id = req.params[ID];
+    const boData: BeneficialOwnerOther = getFromApplicationData(req, BeneficialOwnerOtherKey, id, true);
+
+    let trustIds: string[] = [];
```
Comment (lduranteau) [inline src/utils/beneficial.owner.other.ts:110]: Same as above

File: src/utils/beneficial.owner.other.ts [hunk 8]
```diff
@@ -104,11 +104,23 @@ export const updateBeneficialOwnerOther = async (req: Request, res: Response, ne
...
     // Remove old Beneficial Owner
-    removeFromApplicationData(req, BeneficialOwnerOtherKey, req.params[ID]);
+    removeFromApplicationData(req, BeneficialOwnerOtherKey, id);
 
     // Set Beneficial Owner data
-    const data: ApplicationDataType = setBeneficialOwnerData(req.body, req.params[ID]);
+    const data: ApplicationDataType = setBeneficialOwnerData(req.body, id);
+
+    if (trustIds.length > 0) {
+      (data as BeneficialOwnerOther).trust_ids = trustIds;
```
Comment (lduranteau) [inline src/utils/beneficial.owner.other.ts:122]: Same as above

File: src/controllers/update/update.review.beneficial.owner.individual.ts (hunk 1 above)
Comment (mwestacott) [reply src/controllers/update/update.review.beneficial.owner.individual.ts]: Thanks. Have implemented this now.

File: src/controllers/update/update.review.beneficial.owner.individual.ts (hunk 2 above)
Comment (mwestacott) [reply src/controllers/update/update.review.beneficial.owner.individual.ts]: Done.

**Review by lduranteau — APPROVED** [review]

**Pull request merged**